fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/particles/ParticleFactoryRegistry.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.particles;
+
+import net.fabricmc.fabric.impl.particles.ParticleFactoryRegistryImpl;
+import net.minecraft.client.particle.ParticleFactory;
+import net.minecraft.particle.ParticleEffect;
+import net.minecraft.particle.ParticleType;
+
+/**
+ * Registry for adding particle factories on the client.
+ */
+public interface ParticleFactoryRegistry {
+	static ParticleFactoryRegistry getInstance() {
+		return ParticleFactoryRegistryImpl.INSTANCE;
+	}
+
+	/**
+	 * Registers a factory for constructing particles of the given type.
+	 */
+	<T extends ParticleEffect> void register(ParticleType<T> type, ParticleFactory<T> factory);
+
+	/**
+	 * Registers a delayed factory for constructing particles of the given type.
+	 *
+	 * The factory method will be called with a sprite provider to use for that particle when it comes time.
+	 *
+	 * Particle sprites will be loaded from domain:/particles/particle_name.json as per vanilla minecraft behaviour.
+	 */
+	<T extends ParticleEffect> void register(ParticleType<T> type, PendingParticleFactory<T> constructor);
+
+	/**
+	 * A pending particle factory.
+	 *
+	 * @param <T> The type of particle effects this factory deals with.
+	 */
+	@FunctionalInterface
+	public interface PendingParticleFactory<T extends ParticleEffect> {
```

Comment:
No need for extra public

Reply:
I wasn't sure if it was needed or not, so I left it to be safe. Also doesn't Fabric have some requirement for `public` on interface members?

Reply:
Should be consistent, either put the redundant public modifier everywhere or nowhere. The codebase tends towards the former.

Reply:
Alright. Should I update my interfaces to add the redundant modifier then?

I don't remember if this has always been the case, or if the modifier being optional is new to java 8+.

Reply:
Updated to include the public for all the interfaces. My IDE removes all redundant modifiers on save so it was... difficult.

Reply:
Also an interesting observation: My IDE deems all the `public`s as redundant in interfaces, except the ones on the inner-class definitions of interfaces.